truss-chains/truss_chains/framework.py

```diff
@@ -134,96 +243,116 @@ def _instantiation_error_msg(cls_name: str):
+        f"Unsupported I/O type for `{param_name}` of type `{annotation}`. "
+        "Supported are:\n"
+        f"\t* simple types: {types_str}\n"
+        "\t* containers of these simple types, with annotated item types: "
+        f"{containers_str}, e.g. `dict[str, int]` (use built-in types, not "
+        "`typing.Dict`).\n"
+        "\t* For complicated / nested data structures: `pydantic` models."
+    )
     if isinstance(annotation, types.GenericAlias):
         if get_origin(annotation) not in _SIMPLE_CONTAINERS:
-            raise definitions.ChainsUsageError(error_msg)
+            _collect_error(error_msg, _ErrorKind.IO_TYPE_ERROR, location)
+            return
         args = get_args(annotation)
         for arg in args:
             if arg not in _SIMPLE_TYPES:
-                raise definitions.ChainsUsageError(error_msg)
+                _collect_error(error_msg, _ErrorKind.IO_TYPE_ERROR, location)
+                return
         return
     if utils.issubclass_safe(annotation, pydantic.BaseModel):
         return
 
-    raise definitions.ChainsUsageError(error_msg)
+    _collect_error(error_msg, _ErrorKind.IO_TYPE_ERROR, location)
 
 
 def _validate_endpoint_params(
-    params: list[inspect.Parameter], cls_name: str
+    params: list[inspect.Parameter], location: _ErrorLocation
 ) -> list[definitions.InputArg]:
     if len(params) == 0:
-        raise definitions.ChainsUsageError(
-            f"`{cls_name}.{definitions.ENDPOINT_METHOD_NAME}` must be a method, i.e. "
-            f"with `{definitions.SELF_ARG_NAME}` argument."
+        _collect_error(
+            f"`Endpoint must be a method, i.e. with `{definitions.SELF_ARG_NAME}` as "
+            "first argument. Got function with no arguments.",
+            _ErrorKind.TYPE_ERROR,
+            location,
         )
+        return []
     if params[0].name != definitions.SELF_ARG_NAME:
-        raise definitions.ChainsUsageError(
-            f"`{cls_name}.{definitions.ENDPOINT_METHOD_NAME}` must be a method, i.e. "
-            f"with `{definitions.SELF_ARG_NAME}` argument."
+        _collect_error(
+            f"`Endpoint must be a method, i.e. with `{definitions.SELF_ARG_NAME}` as "
+            f"first argument. Got `{params[0].name}` as first argument.",
+            _ErrorKind.TYPE_ERROR,
+            location,
         )
     input_args = []
     for param in params[1:]:  # Skip self argument.
         if param.annotation == inspect.Parameter.empty:
-            raise definitions.ChainsUsageError(
-                "Inputs of endpoints must have type annotations. For "
-                f"`{cls_name}.{definitions.ENDPOINT_METHOD_NAME}` parameter "
-                f"`{param.name}` has no type annotation."
+            _collect_error(
+                "Arguments of endpoints must have type annotations. "
+                f"Parameter `{param.name}` has no type annotation.",
+                _ErrorKind.IO_TYPE_ERROR,
+                location,
             )
```

Comment:
Should we break here?

Reply:
The loop body is if/else, so below coe won't run in that case. But we can still collect more validation errors from the other params in the param list.